### PR TITLE
JWWEB-620: Initial router5 setup

### PR DESCRIPTION
**Pull request by stcherenkov**
# Core concepts

1. Routing state machine happens before business-logic, so routing is independent layer
2. Description of routing states and ability to move from one to another is gathered in 1 file so it could be observed all at once
3. There is difference between path and name. Path is public interface for browser (or user) to tell us they want to move state machine into specific state. Name is internal API that tells how to move throw state machine and what part of it we should check out
4. Routing layer has 2 sets of API – redux-like and react-like, redux-like takes actions, react-like requires links (see JLink as try to simplify access to that API)

### CODE REVIEW

File: src/components/base/JLink/JLink.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  const { href } = initialProps
+  const props = omit(initialProps, [
+    'theme',
+    'className',
+    'activeClassName',
+    'rel',
+  ])
+  const className = classnames(
+    initialProps.theme && jLinkStyle[initialProps.theme],
+    initialProps.className,
```
Comment (IViol) [inline src/components/base/JLink/JLink.js]: would be great to destructure `theme` & `className` props as well as `href`

File: src/components/base/JLink/tests/JLink.test.js [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    )
+
+    expect(wrapper.text()).toBe('some path')
+    expect(wrapper.find('a').length).toBe(1)
+    expect(wrapper.find('a').prop('href')).toBe('/some/path')
+    expect(wrapper.find('a').prop('target')).toBeUndefined()
+  })
+
+  test('calls router on internal link click', () => {
+    const router = createRouter5([
```
Comment (IViol) [inline src/components/base/JLink/tests/JLink.test.js]: it's ok to reuse it

File: src/components/base/JLink/JLink.js (hunk 1 above)
Comment (stcherenkov) [reply src/components/base/JLink/JLink.js]: Fixed

File: src/components/base/JLink/tests/JLink.test.js (hunk 2 above)
Comment (stcherenkov) [reply src/components/base/JLink/tests/JLink.test.js]: Router itself needs to be created each time to guarantee tests independence. But config can be reused – done.

File: src/routes/NotFound/NotFoundViewContainer.js [hunk 3]
```diff
@@ -2,14 +2,8 @@
...
-
 import NotFoundView from './NotFoundView'
 
-const mapDispatchToProps = {
-  goToHome,
-}
-
 export default (
-  connect/* :: < AppState, any, OwnPropsEmpty, _, _ > */(null, mapDispatchToProps)
+  connect/* :: < AppState, any, OwnPropsEmpty, _, _ > */(null, null)
```
Comment (IViol) [inline src/routes/NotFound/NotFoundViewContainer.js]: so, `connect` could be removed

Comment (stcherenkov) [reply src/routes/NotFound/NotFoundViewContainer.js]: fixed

File: flow-typed/redux/appState.js [hunk 4]
```diff
@@ -7,6 +7,19 @@ import {
   type Channel,
 } from 'redux-saga'
 
+declare type Router5Route = {
+  +name: string,
+  +params: {
+    [key: string]: any,
+  },
+  +path: string,
+}
```
Comment (Creepypoke) [inline flow-typed/redux/appState.js]: Please, add `TODO` for `meta` property pass

File: src/components/AssetItem/AssetItem.js [hunk 5]
```diff
@@ -15,17 +15,10 @@ import {
...
 
 import {
-  JCard,
-  JIcon,
-  JText,
-  JSwitch,
-  JTooltip,
-  JAssetSymbol,
-  JFlatButton,
+  JCard, JIcon, JText, JSwitch, JTooltip, JAssetSymbol, JFlatButton, JLink,
```
Comment (Creepypoke) [inline src/components/AssetItem/AssetItem.js]: ESlint error

File: src/components/base/JLink/JLink.js [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  checkIsExternal,
+} from 'utils/uri'
+
+import jLinkStyle from './JLink.m.scss'
+import { JLinkInternal } from './JLinkInternal'
+
+export type Theme = 'text-white' | 'text-blue'
+
+// base component with inexact props
+export type JLinkProps = {
```
Comment (Creepypoke) [inline src/components/base/JLink/JLink.js]: Looks like typing should be `StyledComponent<Theme> & { ... }`?

Comment (stcherenkov) [reply src/components/base/JLink/JLink.js]: Does not work for me, I need this component to be available completely without styles (meaning – without theme).

File: src/components/AssetItem/AssetItem.js (hunk 5 above)
Comment (stcherenkov) [reply src/components/AssetItem/AssetItem.js]: Fixed in related PR #719

File: flow-typed/redux/appState.js (hunk 4 above)
Comment (stcherenkov) [reply flow-typed/redux/appState.js]: Fixed in related PR #719

**Pull request merged**